### PR TITLE
Fix Java runtime toolchain resolution in cross-compilation scenarios

### LINKED ISSUE #17085: Bazel 6 Remote JDK Toolchains Incorrect Configuration
### Description of the bug:

Upgrading to Bazel 6.0.0, I get some toolchain resolution and build failures around java_library/java_binary rules.  My environment is cross-compile heavy and has remote executors on a number of platforms (linux, darwin both x86 and arm64, windows) and bazel itself is invoked on any of those platforms building for any other.

I get issues when, due to peculiarities in execution platform selection order, an exec platform is chosen for java that is distinct from the target platform.  For instance, target platform is macos_arm64, but the exec platform is @local_config_platform//:host on linux, or a remote linux platform.  For instance, here is toolchain_resolution_debug from a cross compile:

```
INFO: ToolchainResolution:   Type @bazel_tools//tools/jdk:runtime_toolchain_type: target platform //:macos_arm: execution //:macos_arm: Selected toolchain @remotejdk11_macos_aarch64//:jdk
INFO: ToolchainResolution:   Type @bazel_tools//tools/jdk:runtime_toolchain_type: target platform //:macos_arm: execution //:macos: Selected toolchain @remotejdk11_macos_aarch64//:jdk
INFO: ToolchainResolution:   Type @bazel_tools//tools/jdk:runtime_toolchain_type: target platform //:macos_arm: execution //:linux: Selected toolchain @remotejdk11_macos_aarch64//:jdk
INFO: ToolchainResolution:   Type @bazel_tools//tools/jdk:runtime_toolchain_type: target platform //:macos_arm: execution //test:downloader: Selected toolchain @remotejdk11_macos_aarch64//:jdk
INFO: ToolchainResolution:   Type @bazel_tools//tools/jdk:runtime_toolchain_type: target platform //:macos_arm: execution @local_config_platform//:host: Selected toolchain @remotejdk11_macos_aarch64//:jdk
```

A number of these selections are wrong: remotejdk11_macos_aarch64 will not run on the host platform (it is linux), nor will it run on the //:linux remote execution platform.  Nor will it run on the //test:downloader platform which specifies no OS constraints to match at all.  If bazel happens to choose any but the first configuration with identical target and exec platform, it will try to execute a darwin arm64 binary on linux with predictable fireworks.  In the ultimate case where the target and set of exec platforms are distinct, this is guaranteed to fail.  

This also fails when targeting a platform that has no JDK such as Android.  In Bazel 5.1 it worked just fine to have a java_library depended on by an android_library, now toolchain resolution just fails if you set --platforms to one with @platforms//os:android because there is no openjdk for android (and if there were, my remote exec environment has no android runners)

It seems that d5559c16ac008b86345fbbade5d600181a2fce6f changed the definition of the java toolchains to specify only target_compatible_with, not exec_compatible_with, even though the jdk itself is platform-specific.  I'm not sure I understand the benefits of that change, but at a minimum it seems like exec_compatible_with should _also_ be specified for those toolchains.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

See https://github.com/jlaxson/bazel_java_toolchain_bug for the simple reduction involving android platforms.

### Which operating system are you running Bazel on?

Macos, linux

### What is the output of `bazel info release`?

release 6.0.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### LINKED ISSUE #18265: Bazel treats local_jdk as a "can run anywhere" toolchain
### Description of the bug:

I have a shell test which tries to run Java, and fetches it by adding `"@bazel_tools//tools/jdk:current_java_runtime"` to both `toolchains` and `data`.

If I have not configured any Java toolchains, and try to run this test from my host (which is a Mac) remotely on an Linux remote execution worker (exec=target=linux), rather than failing because no toolchain could be located, it brings along the `local_jdk` toolchain, which is a Mac toolchain.

This appears to be because the `local_java_repository` doesn't constraint the `toolchain` target it generates at all: https://cs.opensource.google/bazel/bazel/+/e8cedf459936895f939906474d9f9878a8a45edc:tools/jdk/local_java_repository.bzl;l=91-96

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

BUILD file:
```starlark
sh_test(
    name = "list_test",
    srcs = ["list_test.sh"],
    env = {
        "JAVA": "$(JAVA)",
    },
    toolchains = ["@bazel_tools//tools/jdk:current_java_runtime"],
    data = ["@bazel_tools//tools/jdk:current_java_runtime"],
)
```

list_test.sh:
```
#!/bin/bash -eux

if [[ "${JAVA}" == external/* ]]; then
  JAVA="${JAVA:9}"
fi
"${RUNFILES_DIR}/${JAVA}"
```

Command run on a Mac:
```console
% bazel test :list_test --platforms=@some_linux_platform --extra_execution_platforms=@some_linux_platform --remote_executor=grpcs://linux-remote-execution-server
```

If I run locally on my Mac, I get the output of java help printed, if I run remotely, I get:
```
==================== Test output for //:list_test:
+ [[ external/local_jdk/bin/java == external/* ]]
+ JAVA=local_jdk/bin/java
+ /runner/build/4fde3789d4c6a2f3/root/bazel-out/darwin-fastbuild/bin/list_test.runfiles/local_jdk/bin/java
/runner/build/4fde3789d4c6a2f3/root/bazel-out/darwin-fastbuild/bin/list_test.runfiles/__main__/list_test: line 6: /runner/build/4fde3789d4c6a2f3/root/bazel-out/darwin-fastbuild/bin/list_test.runfiles/local_jdk/bin/java: cannot execute binary file: Exec format error
```

### Which operating system are you running Bazel on?

macOS

### What is the output of `bazel info release`?

release 6.1.1

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

/cc @fmeum who was helping me debug, and has https://github.com/bazelbuild/bazel/pull/18262 as a possible solution

### PR DESCRIPTION
**Pull request by fmeum**
The Java rules used toolchains of type `@bazel_tools//tools/jdk:runtime_toolchain_type` for two different purposes requiring different exec/target platform constraints, which led to issues when cross-compiling: The runtime added to the runfiles of an executable Java target has to have constraints on the target platform, whereas the runtime used to extract the bootclasspath should not have constraints on the target platform. As a result:
1. `@local_jdk` did not define any target constraints for its runtime, which allowed the runtime to provide the bootclasspath required by Android rules, but could also end up building `java_binary` targets that wouldn't run on the target (see #18265).
2. Remote JDKs defined a target constraint for the their runtimes, which prevented them from being added to the runfiles of targets they couldn't run on, but broke Android compilation due to the failure to resolve a runtime for bootclasspath extraction (see #17085).

This is resolved by adding a third toolchain type, `bootstrap_runtime_toolchain_type`, that is only used by the `bootclasspath` rule (realizes https://github.com/bazelbuild/bazel/issues/17085#issuecomment-1454306360). Detailed explanations of the different types and their intended constraints have been added to `@bazel_tools//tools/jdk:BUILD`.

Addressing the cross-compilation errors then required the following changes:
1. Direct dependencies on the Java runtime have been removed from rules that do not need them (e.g. `android_library` and `java_binary` with `create_executable = False`).
3. Implicit dependencies on the Java runtime through the `bootclasspath` rule have been replaced with dependencies on the bootstrap runtime.
4. `{local,remote}_java_repository` now use the user-provided exec constraints of the Java toolchain as the target constraints of the associated runtime. 
5. `@local_jdk` uses the auto-detected host constraints as exec constraints for the local Java toolchain.

Fixes #17085
Fixes #18265 
Fixes bazelbuild/rules_java#64

RELNOTES[INC]: Java runtime toolchains created via `local_java_repository` from `@bazel_tools//tools/jdk:local_java_repository.bzl`, which includes `local_jdk`, now have `target_compatible_with` set to the auto-detected host constraints. This can result in errors about toolchain resolution failures for `@bazel_tools//tools/jdk:runtime_toolchain_type`, especially when cross-compiling. These failures can be fixed in the following ways (listed in decreasing order of preference):
* Replace `java_binary` targets that aren't meant to be run with `bazel run` or as tools during the build with `java_single_jar` (available in `@rules_java//java:java_single_jar.bzl`). Such targets do not require a Java runtime for the target configuration.
* Set `--java_runtime_version=remotejdk_N` for some Java version `N` to let Bazel choose and download an appropriate remote JDK for the current target platform. This setting defaults to `local_jdk`, which means that Bazel can only use the local JDK, which isn't compatible with any other platform.
* Manually define and register a `local_java_runtime` with no value set for `exec_compatible_with` (defaults to `[]`) and select it by setting `--java_runtime_version` to its `name`. This fully restores the previous behavior, but can result in incorrect results when cross-compiling (see #18265).

### CODE REVIEW

**Comment by fmeum** [comment]
Looks like this is running into https://github.com/bazelbuild/bazel/issues/17085 now.

**Comment by fmeum** [comment]
@cushon Could you review the first three commits of this PR? They implement what you described in https://github.com/bazelbuild/bazel/issues/17085#issuecomment-1454306360.

@hvadehra Could you review the last commit as it touches the Starlark version of `java_binary`? Since I can't make changes to Google's `java_semantics.bzl`, I added a placeholder files with the changes required to make that commit a noop for Google.

Please let me know if you prefer having this PR split into multiple ones. I tried to group independently reviewable changes into separate commits, but only the combination of these commits truly resolves the cross-compilation issues.

FYI @illicitonion @jlaxson Please test with your setups if possible.

**Comment by illicitonion** [comment]
I tested this out and it worked for me - I got a useful "no toolchain found" error when I didn't have one registered, and when I registered one, everything worked as I expected. Thanks so much!

File: tools/jdk/BUILD.tools
Comment (cushon) [inline tools/jdk/BUILD.tools:51]: This is a nice detail.

For Google's internal configuration I have currently landed on just using a `select()` on `--java_language_version=` to get a JDK to use as the bootclasspath, because as you point out here it's platform independent. But for Bazel it's nice to be able to use something that's already fetched, if possible.

**Review by cushon — COMMENTED** [review]
I'm on board with the overall approach here, thanks for tackling this.

I would like @hvadehra's input as well.

I think it probably makes sense to split some of the commits into separate PRs, maybe after everyone's taken a first pass.

**Comment by hvadehra** [comment]
LGTM in general. 

+1 to splitting this into separate commits. In particular, I think I'd like to do the changes in semantics.bzl slightly differently. But it's okay to leave as is on github, I'll handle that while importing.

**Comment by fmeum** [comment]
I will wait for https://github.com/bazelbuild/rules_java/pull/110 to land before I continue working on this.

**Comment by fmeum** [comment]
I split off four PRs:

* https://github.com/bazelbuild/bazel/pull/18618
* https://github.com/bazelbuild/bazel/pull/18620
* https://github.com/bazelbuild/rules_java/pull/113
* https://github.com/bazelbuild/rules_java/pull/114 (draft, need to clarify some changes caused by the recent move of toolchains to rules_java)

I will keep this one open and change it to contain only the rules_java update and tests after all the other PRs have been merged.

**Comment by fmeum** [comment]
@hvadehra I updated the PR to reference the new rules_java version.

Most tests are passing, but I don't really know what to do about the failing ones:
* `test_java_runtime` is failing because the custom local JDK is selected for compilation based on `--tool_java_runtime_version`. I would have expected that to be correct: Shouldn't the runtime used for compilation be configured based on `--tool_java_runtime_version`, not `--java_runtime_version`? The test does appear to have this the other way round.
* RBE tests are failing since the RBE toolchain doesn't register the bootstrap runtime toolchain. Could you point me to the toolchain definitions so that I can update them?

**Comment by hvadehra** [comment]
Sorry, didn't get a chance, will look into this tomorrow.

**Comment by hvadehra** [comment]
re: RBE, it looks like it's fetched from `https://storage.googleapis.com/bazel-ci/rbe-configs/manifest.json` (we pick the first entry if we don't get a match for current bazel version). We appear to be using [ubuntu1804-bazel-java11](https://github.com/bazelbuild/bazel/blob/00dfa2ee8cce6722847c60bfb0626458503741af/WORKSPACE#L160). 

If you unpack the tar you can see the toolchain(s) defined. 

`java/BUILD` looks like:

```
# This file is auto-generated by github.com/bazelbuild/bazel-toolchains/pkg/rbeconfigsgen
# and should not be modified directly.

load("@bazel_tools//tools/jdk:local_java_repository.bzl", "local_java_runtime")

package(default_visibility = ["//visibility:public"])

alias(
    name = "jdk",
    actual = "rbe_jdk",
)

local_java_runtime(
    name = "rbe_jdk",
    java_home = "/usr/lib/jvm/java-11-openjdk-amd64",
    version = "11.0.11",
)
```

So I would think the bootstrap toolchain should have gotten declared by `local_java_runtime`?

But this is all new to me, so @meteorcloudy please confirm if this sounds right?

**Comment by fmeum** [comment]
Thanks for the pointers, this was very helpful.

The RBE definitions end up defining a `local_java_runtime` coming from `@bazel_tools//tools/jdk:local_java_repository.bzl`, which in turn, via repo mappings, uses the `@rules_java_builtin` definition of that macro. Crucially, this is the `rules_java` shipped with the version of Bazel that performs the build, *not* the version of Bazel being built. 

In the CI situation, this means that the Java toolchains come from a pre-6.5.0 version of rules_java, which doesn't register the bootstrap runtime toolchain. Since we restrict the `--java_runtime_version` to `rbe_jdk`, the 6.5.0 versions of the toolchains with the bootstrap runtime registered in `WORKSPACE` are ignored.

But the Java compilation toolchains from `rules_java`, which take precedence over those registered for `rules_java_builtin` in the WORKSPACE suffix, [reference `@rules_java//toolchains:platformclasspath` directly](https://github.com/bazelbuild/rules_java/blob/master/toolchains/default_java_toolchain.bzl#L89), which in turn has a dependency on a bootstrap runtime toolchain, which then isn't found among the `rbe_jdk` runtimes.

We could replace the reference to the `platformclasspath` rule in `rules_java` with one pointing to `@bazel_tools`. I will see whether that works.

@hvadehra Do you have other ideas?

**Comment by fmeum** [comment]
That didn't work. I don't know why anything would be triggering `@rules_java//toolchains:platformclasspath` to be built as all remaining references should go through `@bazel_tools` and thus point to `@rules_java_builtin//toolchains:platformclasspath` due to repo mapping.

@meteorcloudy You are needed here :-)

**Comment by hvadehra** [comment]
Hang on, ISTM we've ended up in a situation, that is extremely surprising at best, perhaps incorrect at worst.

Considering `@bazel_tools//tools/jdk:local_java_repository.bzl` only exists for backwards-compatibility reasons, the more principled way forward might be to update the rbe toolchains to use `@rules_java` directly. @meteorcloudy WDYT?

**Comment by hvadehra** [comment]
Alternatively, would it work if we wait for a Bazel version[^1] that ships will an embedded rules_java >= 6.5.0?

[^1]: maybe 6.4.0 ?

**Comment by fmeum** [comment]
I think that raising that embedded version would require exactly the changes in the current PR ;-) We do have a bootstrapping problem.

**Comment by hvadehra** [comment]
> I think that raising that embedded version would require exactly the changes in the current PR

Do you mean to say that updating to `rules_java` to 6.5.0 is breaking and cannot be done _before_ the other changes in this PR?

**Comment by fmeum** [comment]
Summarizing what I found:
* rules_java 6.5.0 requires a host Bazel containing https://github.com/bazelbuild/bazel/commit/8715e9ae6f5c2fd42607018b3adaad14bf601a1c, which is why I bumped `.bazelversion` to a pre-release. This commit will be cherry-picked into Bazel 6.4.0, which we could use once it has been released.
* rules_java without https://github.com/bazelbuild/rules_java/pull/126 also requires a Bazel that embeds rules_java 6.5.0 or later, which creates a circular dependency that I don't think we can resolve without temporarily disabling certain CI checks. 
* rules_java with https://github.com/bazelbuild/rules_java/pull/126 should allow updating the rules_java pin Bazel to that version (say 6.5.1).

**Comment by hvadehra** [comment]
re: failing test cases in `//src/test/shell/integration:bazel_java_test`, see https://github.com/bazelbuild/bazel/issues/17281#issuecomment-1399608274

**Comment by hvadehra** [comment]
@fmeum I think I have a way forward for the RBE failures. Please stay tuned :)

**Comment by fmeum** [comment]
@hvadehra That's great, I think I finally figured out the other test failure thanks to your pointer. Submitted https://github.com/bazelbuild/rules_java/pull/127 to fix the docs. With https://github.com/bazelbuild/rules_java/pull/126 patched in, this should be working now :-)

**Comment by fmeum** [comment]
@meteorcloudy Could you run a downstream test for this PR, assuming it's not too red at HEAD?

**Comment by meteorcloudy** [comment]
Sure, https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/3247

**Comment by hvadehra** [comment]
@fmeum Can you try dropping 360900b46aedba13568185d181e5f7e88cac67c0, I think it RBE should work now.

**Comment by fmeum** [comment]
@hvadehra Looks like all tests are passing with it. Could you explain what you changed? I am worried about this coming up in other contexts that CI doesn't cover.

**Comment by hvadehra** [comment]
Haha, so I updated https://storage.googleapis.com/bazel-ci/rbe-configs/manifest.json (see bottom of the file for the new config), which for the current rolling release uses `@rules_java` directly instead of `@bazel_tools`. 

This is somewhat temporary, and will break your changes when we have a new rolling release. The proper fix will land after:
 
https://github.com/bazelbuild/bazel-toolchains/pull/999 
https://github.com/bazelbuild/continuous-integration/pull/1726

and then of course making a new release of `bazel_ci_rules` and updating bazel to use it.

**Comment by fmeum** [comment]
I think that it may be too early to instruct users to update their Java toolchain registrations to rules_java: When the `@rules_java` version is not at least 6.5.0, the runtimes registered from it will lack the bootstrap runtime toolchain registration. However, the default compilation toolchain always comes from `@bazel_tools`'s `@rules_java_builtin`, which for a Bazel including this PR always uses a `platformclasspath` that relies on the bootstrap runtime toolchain. Getting this mixed up results in very baffling error messages.

Accidentally pulling in old versions of `@rules_java` happens frequently due to WORKSPACE's "first declaration wins, except for surprising exceptions to this rule" semantics. Sorting out Java toolchain registration will already cause some disruption (see the downstream run, which shows that Bazel itself relies on local_jdk being target compatible with everything for Windows arm64 builds), but I think that's unavoidable. We can however simplify the migration for users if we keep suggesting `@bazel_tools` and merge bazelbuild/rules_java#126 instead of making these changes to the RBE toolchains.

(I actually think that we won't be able to fully separate rules_java from Bazel without relying on Bzlmod for as long as rules_java is still somewhat tightly coupled to specific Bazel versions, but that's a broader topic.)

**Comment by hvadehra** [comment]
Ack, that makes sense. 

I'll import https://github.com/bazelbuild/rules_java/pull/126 and release 6.5.1*

* perhaps we can start declaring the toolchain types in rules_java and make the next release 7.0.0

**Comment by hvadehra** [comment]
@fmeum Do you have more context on what exactly is going wrong with windows arm64?

**Comment by fmeum** [comment]
There is also bazelbuild/rules_java#127, which is a doc-only fix but would ideally also go into the next release.

**Comment by fmeum** [comment]
> @fmeum Do you have more context on what exactly is going wrong with windows arm64?

The Windows arm64 build cross-compiles, which means that `@local_jdk` (a Windows x86_64 JDK) isn't target compatible with the target platform. See https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/17094#018a1c4d-e641-4a77-8a75-0f6d62495ca6, which has more log output.

This is a potential breaking change caused by this PR: Cross-platform builds of a `java_binary` would previously silently package a `@local_jdk` runtime, which of course won't run on the target. This was clearly a bug, but with this fixed such builds require a bit more care. In this case, we could either:
* set `--java_runtime_version=remotejdk` so that an appropriate remote JDK can be downloaded or
* mark the Bazel server `java_binary` target with `create_executable = False`.

**Comment by hvadehra** [comment]
Thanks!

From what I can tell, the Bazel server is used as an executable, so we should go with option 1. And IIUC, after this PR, that is the way forward for cross-compile java builds?

**Comment by fmeum** [comment]
> From what I can tell, the Bazel server is used as an executable, so we should go with option 1. 

Okay, I will submit a separate PR to do this.

> And IIUC, after this PR, that is the way forward for cross-compile java builds?

Could you elaborate on that question? For any other project, there is also the same choice between 1. and 2. above. Alternatively, the old behavior can be restored by registering a local Java repository with no constraints. 

We should document this in RELNOTES.

**Comment by hvadehra** [comment]
Thanks, that answers my question. 

Good call on the release notes. Could you please add a `REL-NOTES[INC]: blah blah` to the PR description?

**Comment by hvadehra** [comment]
rules_java 6.5.1 released: https://github.com/bazelbuild/rules_java/releases/tag/6.5.1

**Comment by fmeum** [comment]
@hvadehra Thanks! 

I am trying to switch over to option 1. as 2. caused failures in the integration test setup that I did not know how to resolve. Will see how that goes.

**Comment by fmeum** [comment]
CI is still failing since https://github.com/bazelbuild/bazel/commit/132d2497f4337473fb084677c7f2b98c85e67c92 is not part of the Bazel listed in `.bazelversion`. We may need to wait for the next rolling release.

**Comment by meteorcloudy** [comment]
Wait, we probably shouldn't change `.bazelversion` after all? Is it just for some testing?

**Comment by meteorcloudy** [comment]
Yes, in general we should build Bazel with the latest LTS release. I thought @hvadehra is working on a solution that will allow  Bazel 6.3.2 to work?

**Comment by fmeum** [comment]
> Yes, in general we should build Bazel with the latest LTS release. I thought @hvadehra is working on a solution that will allow Bazel 6.3.2 to work?

I can look into alternatives, but we will probably need 6.4.0.

**Comment by fmeum** [comment]
I found a simple way to make the arm build work by using a remote JDK for it. I temporarily set `.bazelversion` to a commit on the `release-6.4.0` branch. Once 6.4.0 is out, I can update to that and we can merge this PR.

**Comment by hvadehra** [comment]
Nice, sounds good, thanks!

re: release notes, 

> Mark java_binary targets that aren't meant to be run with bazel run or as tools during the build with create_executable = False. Such targets no longer require a Java runtime for the target configuration.

I think @comius might be a little unhappy with this being the top recommendation :)

**Comment by fmeum** [comment]
> > Mark java_binary targets that aren't meant to be run with bazel run or as tools during the build with create_executable = False. Such targets no longer require a Java runtime for the target configuration.
> I think @comius might be a little unhappy with this being the top recommendation :)

Happy to reword or downgrade that recommendation :-) 

In the projects I maintain, I found that there are plenty of usages of `java_binary` that are only meant to be used as `deploy_env`s for other targets or for their `_deploy.jar`s. For these targets, I find it both more readable and more correct to label them as non-executable. Does that sound more agreeable?

**Comment by fmeum** [comment]
@hvadehra I stacked this onto #19793 to get the tests to pass after the flip of `--incompatible_enable_cc_toolchain_resolution`. I also updated the PR description with the relnotes now recommending `java_single_jar`. Once a new rules_java release with that rule has been cut, I could update this PR to point to it.

**Comment by fmeum** [comment]
@katre I think that this PR fails in CI since https://github.com/bazelbuild/bazel/commit/5efeaa5a8562f8517917c875c93a05c5fc6a72b2#diff-0ee346b19181339d306c6c5dea526e48eb13ef61e93fe46dae8281a7d5691451, which sets `--platforms` to an Android platform for `android_local_test` tests, which results in a Java runtime being searched for Android arm7v. Before the current PR, the `local_jdk` doesn't have any `target_compatible_with` constraints and thus happily claims to run on Android, which it clearly doesn't. With that fixed, the `android_local_test` tests now fail since they can't find a runtime.

Since an `android_local_test` is supposed to run on the local machine, is it really valid to invoke it with `--platforms` set to an Android platform? If that needs to work for some reason, the Java runtime dependency should probably go through a transition forcing it back to e.g. the host platform.

**Comment by katre** [comment]
Okay, this is exciting.

The rule `android_local_test` wants to do two things: it wants to [build an APK](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/android/AndroidLocalTestBase.java;drc=196515539edb352f10835f38aeea4d7efa832643;l=99), using a [valid Android SDK](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/rules/android/BazelAndroidLocalTestRule.java;drc=5efeaa5a8562f8517917c875c93a05c5fc6a72b2;l=104).

But, it also wants to run a Java binary on the local platform, with that APK as a data file. This sounds like a case for execution groups to me, but I also don't actually understand how this works under the hood.

I'm filing an issue to track updating `android_local_test` and then disabling the test so you can proceed.

**Comment by fmeum** [comment]
@hvadehra Tests are passing now, but this requires https://github.com/bazelbuild/bazel/issues/19837 to be resolved first, which in turn requires https://github.com/bazelbuild/rules_java/pull/142.

**Comment by hvadehra** [comment]
@fmeum Could you please change this to use rules_java 7.0.0?

**Comment by fmeum** [comment]
@hvadehra I updated to rules_java 7.0.0 and also included a small workaround for #19837 for Bazel itself.

**Comment by hvadehra** [comment]
@fmeum you might need some of the changes from #19864 (especially 74f8caa76632c511c5f2921981ad26c27150f7fe)

(i rebased that on top of the earlier version of this, and things look good, so if it's easier, we could go with the alternative and merge that, and then proceed with #19864 after)

**Comment by fmeum** [comment]
> @fmeum you might need some of the changes from #19864 (especially [74f8caa](https://github.com/bazelbuild/bazel/commit/74f8caa76632c511c5f2921981ad26c27150f7fe))

I added the changes here, but I think that the failure is legitimate: According to https://github.com/bazelbuild/rules_java/blob/0c538b8720a8b8df5f50ce92debffc8b27b84af4/toolchains/BUILD#L260-L315, rules_java will use the following JDKs to compile for the different target Java versions:

```
8, 9, 10, 11, 21: JDK 21
14: JDK 14
15: JDK 15
16: JDK 16
17: JDK 17
```

This is also what we are seeing in the test: `--java_language_version=17` results in compilation with JDK 17, not JDK 21.

@hvadehra Should we change the linked part of the `BUILD` file to something like this?

```
RELEASES = (8, 9, 10, 11, 17, 21)

[
    default_java_toolchain(
        name = ("toolchain_java%d" if release <= 11 else "toolchain_jdk_%d") % release,
        configuration = DEFAULT_TOOLCHAIN_CONFIGURATION,
        source_version = "%s" % release,
        target_version = "%s" % release,
    )
    for release in RELEASES
]
```

I think that we can drop JDK 14, 15, and 16. We could also decide to consolidate the name to `toolchain_jdk_%d`.

**Comment by hvadehra** [comment]
> @hvadehra Should we change the linked part of the BUILD file to something like this?

Yeah, that makes sense, I'll make the change. 

Is the RBE failure due to the attempted workaround for https://github.com/bazelbuild/bazel/issues/19837 ?

**Comment by fmeum** [comment]
> Is the RBE failure due to the attempted workaround for #19837 ?

Good question, no idea yet, but luckily I can debug this locally :-)

**Comment by fmeum** [comment]
This is caused by the `register_toolchains("//:bazel_java_toolchain_definition")` line in `MODULE.bazel`, but I don't understand why.

Without this line:
```
bazel build --nobuild //:bazel_java_toolchain_definition --config=remote # passes
bazel build --nobuild //:bazel_java_toolchain --config=remote # fails with the expected missing toolchain error since this toolchain requires a bootstrap runtime, but none is registered with the rbe_jdk prefix
```

This behavior is expected. But why does `register_toolchains("//:bazel_java_toolchain_definition")` result in the analysis of `//:bazel_java_toolchain_definition`?

**Comment by fmeum** [comment]
If we can't figure this out and can accept the warnings, I can revert the workaround. Unfortunately there doesn't seem to be a way to register a certain toolchain by default, but not with a particular `--config`.

I will look into why this doesn't behave as expected.

**Comment by fmeum** [comment]
Nvm, the RBE toolchains only include Java runtime toolchains, not Java compilation toolchains, resulting in the selection of the `//:bazel_java_toolchain`. I will add another workaround.

**Comment by fmeum** [comment]
@hvadehra Updated

**Comment by fmeum** [comment]
I forgot to update `MODULE.bazel.lock` after running `bazel run //src/test/tools/bzlmod:update_default_lock_file`. I also added a comment about that.

**Comment by fmeum** [comment]
`cc_integration_test` keeps failing, presumably because I haven't updated rules_java in all the right places. I forgot `distdir_deps.bzl`, not sure whether was the last one.

**Comment by hvadehra** [comment]
> not sure whether was the last one.

:crossed_fingers:

**Comment by fmeum** [comment]
Great, I won't touch this anymore.

**Review by hvadehra — APPROVED** [review]

**Review by ahumesky — APPROVED** [review]

**Comment by ahumesky** [comment]
> @ahumesky Do you have any concerns about the change to `src/test/java/com/google/devtools/build/android/dexer/BUILD`? (context is that we're now using jdk21, which does not support java 7)

thanks -- the dx parts are not used anymore and will be removed soon so this looks fine

**Pull request closed without merging**
